include/seqan3/io/alignment_file/output.hpp

```diff
@@ -625,24 +691,31 @@ class alignment_file_output
     //!\brief The actual std::variant holding a pointer to the detected/selected format.
     format_type format;
 
+    //!\brief The header type, which specilised with ref_ids_type if reference information are given.
+    using header_type = alignment_file_header<std::conditional_t<std::Same<ref_ids_type, ref_info_not_given>,
+                                              std::vector<std::string>,
+                                              ref_ids_type>>;
+
+    //!\brief The file header object (will be set on construction).
+    std::unique_ptr<header_type> header_ptr;
+
     //!\brief Write record to format.
-    template <typename ...pack_type>
-    void write_record(alignment_file_header const * hdr_ptr, pack_type && ...remainder)
+    template <typename header_t, typename ...pack_type>
+    void write_record(header_t && hdr_ptr, pack_type && ...remainder)
     {
-        if (header_ptr == nullptr && hdr_ptr != nullptr)
-            header_ptr = std::unique_ptr<alignment_file_header>(new alignment_file_header(*hdr_ptr));
-
         static_assert((sizeof...(pack_type) == 14), "Wrong parameter list passed to write_record.");
 
         assert(!format.valueless_by_exception());
 
         std::visit([&] (auto & f)
         {
-            f.write(stream,
-                    options,
-                    header_ptr,
-                    std::forward<pack_type>(remainder)...);
-
+            // use header from record if explicitly given, e.g. file_out = file_in
+            if constexpr (!std::Same<header_t, std::nullptr_t>)
+                f.write(stream, options, *hdr_ptr, std::forward<pack_type>(remainder)...);
+            else if constexpr (std::Same<ref_ids_type, ref_info_not_given>)
+                f.write(stream, options, std::ignore, std::forward<pack_type>(remainder)...);
+            else
+               f.write(stream, options, *header_ptr, std::forward<pack_type>(remainder)...);
```

Comment:
header_ptr and hdr_ptr are confusing to read here, can you distinguish the names better?